apps/tap/test/tap.js

```diff
@@ -62,64 +85,328 @@ contract('Tap app', accounts => {
-      agentAppId,
-      agentBase.address,
-      '0x',
-      false
-    )
-    const agentProxyAddress = getEvent(agentReceipt, 'NewAppProxy', 'proxy')
-    agent = Agent.at(agentProxyAddress)
+    await acl.createPermission(authorized, tapApp.address, ADD_TOKEN_TAP_ROLE, root, { from: root })
+    await acl.createPermission(authorized, tapApp.address, REMOVE_TOKEN_TAP_ROLE, root, { from: root })
+    await acl.createPermission(authorized, tapApp.address, UPDATE_TOKEN_TAP_ROLE, root, { from: root })
+    await acl.createPermission(authorized, tapApp.address, UPDATE_POOL_ROLE, root, { from: root })
+    await acl.createPermission(authorized, tapApp.address, UPDATE_VAULT_ROLE, root, { from: root })
+    await acl.createPermission(authorized, tapApp.address, WITHDRAW_ROLE, root, { from: root })
+
+    const recoveryVault = await setupRecoveryVault(dao)
+    const initialPool = await setupInitialPool(dao)
+    const agentDispatch = await setupAgentDispatch(dao)
+
+    return { dao, tapApp, recoveryVault, initialPool, agentDispatch }
+  }
+  const forceSendETH = async (to, value) => {
+    // Using this contract ETH will be send by selfdestruct which always succeeds
+    const forceSend = await ForceSendETH.new()
+    return forceSend.sendByDying(to, { value })
+  }
+
+  beforeEach(async () => {
+    const { dao, tapApp, recoveryVault, initialPool, agentDispatch } = await newProxyTap()
+    tap = tapApp
+    pool = initialPool
+    agent = agentDispatch
 
-    await agent.initialize()
+    //vault
+    const receipt1 = await dao.newAppInstance('0x1234', vaultBase.address, '0x', false, { from: root })
+    vault = await Vault.at(getEvent(receipt1, 'NewAppProxy', 'proxy'))
+    const acl = await ACL.at(await dao.acl())
+    await acl.createPermission(tap.address, vault.address, TRANSFER_ROLE, root, { from: root })
+    await vault.initialize()
+    await forceSendETH(pool.address, INITIAL_ETH_BALANCE)
+    token1 = await TokenMock.new(pool.address, 10000)
+
+    await tap.initialize(pool.address, vault.address, INITIAL_TAP_RATE)
+  })
+
+  context("> initialize", () => {
+    it("it should revert on re-initialization", async () => {
+      const newTap = await Tap.new()
+      const newVault = await Vault.new()
+      const newPool = await Pool.new()
+      assert.isTrue(await newTap.isPetrified())
+      assert.isTrue(await newVault.isPetrified())
+      assert.isTrue(await newPool.isPetrified())
+      return assertRevert(async () => {
+        await newTap.initialize(newPool, newVault, 15)
+      })
+    })
+  })
+
+  context("> withdraw", () => {
+    context("ETH", () => {
+      it("it should transfer a tap-defined amount of ETH from the collateral pool to the vault", async () => {
+        await tap.addTokenTap(ETH, tapAmt, { from: authorized })
+        assert.equal(await getBalance(pool.address), INITIAL_ETH_BALANCE, 'pool balance should be correct prior to withdraw')
+        await tap.withdraw(ETH, { from: authorized })
```

Comment:
This line fails on both contexts for ERC20/ETH...
It says that the pool has no tokens to withdraw. It makes sense why ERC20 does not have a token balance since in the `beforeEach` step I only sent ETH to the pool account. Not sure why it's failing in both cases...